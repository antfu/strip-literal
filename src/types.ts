export interface StripLiteralOptions {
  /**
   * Will be called for each string literal. Return false to skip stripping.
   */
  filter?: (s: string) => boolean
}
