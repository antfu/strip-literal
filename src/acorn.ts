import { tokenizer } from 'acorn'

/**
 * Strip literal using Acorn's tokenizer.
 *
 * Will throw error if the input is not valid JavaScript.
 */
export function stripLiteralAcorn(code: string) {
  const FILL = ' '
  let result = ''
  function fulfill(index: number) {
    if (index > result.length)
      result += code.slice(result.length, index).replace(/[^\n]/g, FILL)
  }

  const tokens = tokenizer(code, {
    ecmaVersion: 'latest',
    sourceType: 'module',
    allowHashBang: true,
    allowAwaitOutsideFunction: true,
    allowImportExportEverywhere: true,
  })
  const inter = tokens[Symbol.iterator]()

  while (true) {
    const { done, value: token } = inter.next()
    if (done)
      break
    fulfill(token.start)
    if (token.type.label === 'string')
      result += code[token.start] + FILL.repeat(token.end - token.start - 2) + code[token.end - 1]
    else if (token.type.label === 'template')
      result += FILL.repeat(token.end - token.start)

    else
      result += code.slice(token.start, token.end)
  }

  fulfill(code.length)

  return result
}

/**
 * Returns a function that returns whether the position is
 * in a literal using Acorn's tokenizer.
 *
 * Will throw error if the input is not valid JavaScript.
 */
export function createIsLiteralPositionAcorn(code: string) {
  // literal start position, non-literal start position, literal start position, ...
  const positionList: number[] = []

  const tokens = tokenizer(code, {
    ecmaVersion: 'latest',
    sourceType: 'module',
    allowHashBang: true,
    allowAwaitOutsideFunction: true,
    allowImportExportEverywhere: true,
    onComment(_isBlock, _text, start, end) {
      positionList.push(start)
      positionList.push(end)
    },
  })
  const inter = tokens[Symbol.iterator]()

  while (true) {
    const { done, value: token } = inter.next()
    if (done)
      break
    if (token.type.label === 'string') {
      positionList.push(token.start + 1)
      positionList.push(token.end - 1)
    }
    else if (token.type.label === 'template') {
      positionList.push(token.start)
      positionList.push(token.end)
    }
  }

  return (position: number) => {
    const i = binarySearch(positionList, v => position < v)
    return (i - 1) % 2 === 0
  }
}

function binarySearch(array: ArrayLike<number>, pred: (v: number) => boolean) {
  let low = -1
  let high = array.length
  while (1 + low < high) {
    const mid = low + ((high - low) >> 1)
    if (pred(array[mid]))
      high = mid
    else
      low = mid
  }
  return high
}
