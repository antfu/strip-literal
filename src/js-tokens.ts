import jsTokens, { type Token as JSToken } from 'js-tokens'
import type { StripLiteralOptions } from './types'

export function _stripLiteralJsTokens(code: string, options?: StripLiteralOptions) {
  const FILL = options?.fillChar ?? ' '
  const FILL_COMMENT = ' '
  let result = ''

  const filter = options?.filter ?? (() => true)

  const tokens: JSToken[] = []

  let error: any
  try {
    for (const token of jsTokens(code, { jsx: false })) {
      tokens.push(token)

      if (token.type === 'SingleLineComment') {
        result += FILL_COMMENT.repeat(token.value.length)
        continue
      }

      if (token.type === 'MultiLineComment') {
        result += token.value.replace(/[^\n]/g, FILL_COMMENT)
        continue
      }

      if (token.type === 'StringLiteral') {
        const body = token.value.slice(1, -1)
        if (filter(body)) {
          result += token.value[0] + FILL.repeat(body.length) + token.value[token.value.length - 1]
          continue
        }
      }

      if (token.type === 'NoSubstitutionTemplate') {
        const body = token.value.slice(1, -1)
        if (filter(body)) {
          result += `\`${body.replace(/[^\n]/g, FILL_COMMENT)}\``
          continue
        }
      }

      if (token.type === 'RegularExpressionLiteral') {
        const body = token.value
        if (filter(body)) {
          result += body.replace(/\/(.*)\/(\w?)$/g, (_, $1, $2) => `/${FILL.repeat($1.length)}/${$2}`)
          continue
        }
      }

      // `start${
      if (token.type === 'TemplateHead') {
        const body = token.value.slice(1, -2)
        if (filter(body)) {
          result += `\`${body.replace(/[^\n]/g, FILL_COMMENT)}\${`
          continue
        }
      }

      // }end`
      if (token.type === 'TemplateTail') {
        const body = token.value.slice(0, -2)
        if (filter(body)) {
          result += `}${body.replace(/[^\n]/g, FILL_COMMENT)}\``
          continue
        }
      }

      // }middle${
      if (token.type === 'TemplateMiddle') {
        const body = token.value.slice(1, -2)
        if (filter(body)) {
          result += `}${body.replace(/[^\n]/g, FILL_COMMENT)}\${`
          continue
        }
      }

      result += token.value
    }
  }
  catch (e) {
    error = e
  }

  return {
    error,
    result,
    tokens,
  }
}
