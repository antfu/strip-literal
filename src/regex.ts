import type { StripLiteralOptions } from './types'

const multilineCommentsRE = /\/\*([^*\/])*?\*\//gms
const singlelineCommentsRE = /(?:^|\n|\r)\s*\/\/.*(?:\r|\n|$)/gm
const templateLiteralRE = /\$\{(\s*(?:|{.*}|(?!\$\{).|\n|\r)*?\s*)\}/g
const quotesRE = [
  /(["'`])((?:\\\1|(?!\1)|.|\r)*?)\1/gm,
  /([`])((?:\\\1|(?!\1)|.|\n|\r)*?)\1/gm, // multi-line strings (i.e. template literals only)
]

/**
 * Strip literal using RegExp.
 *
 * This will be faster and can work on non-JavaScript input.
 * But will have some caveats on distinguish strings and comments.
 */
export function stripLiteralRegex(code: string, options?: StripLiteralOptions) {
  const filter = options?.filter ?? (() => true)

  code = code
    .replace(multilineCommentsRE, s => filter(s) ? ' '.repeat(s.length) : s)
    .replace(singlelineCommentsRE, s => filter(s) ? ' '.repeat(s.length) : s)

  let expanded = code
  // Recursively replace ${} to support nested constructs (e.g. ${`${x}`})
  for (let i = 0; i < 16; i++) {
    const before = expanded
    expanded = expanded.replace(templateLiteralRE, '` $1`')
    if (expanded === before)
      break
  }

  quotesRE.forEach((re) => {
    expanded = expanded
      .replace(re, (s, quote, body, index) => {
        if (!filter(s.slice(1, -1)))
          return s
        code = code.slice(0, index + 1) + ' '.repeat(s.length - 2) + code.slice(index + s.length - 1)
        return quote + ' '.repeat(s.length - 2) + quote
      })
  })

  return code
}
