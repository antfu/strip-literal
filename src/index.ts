import { _stripLiteralJsTokens } from './js-tokens'
import { stripLiteralRegex } from './regex'
import type { StripLiteralOptions } from './types'

export { stripLiteralAcorn, createIsLiteralPositionAcorn } from './acorn'
export { stripLiteralRegex } from './regex'
export { _stripLiteralJsTokens } from './js-tokens'
export * from './types'

/**
 * Strip literal from code.
 *
 * Using Acorn's tokenizer first, and fallback to Regex if Acorn fails.
 */
export function stripLiteral(code: string, options?: StripLiteralOptions) {
  return stripLiteralDetailed(code, options).result
}

/**
 * Strip literal from code, return more detailed information.
 *
 * Using Acorn's tokenizer first, and fallback to Regex if Acorn fails.
 */
export function stripLiteralDetailed(code: string, options?: StripLiteralOptions): {
  mode: 'js-tokens' | 'regex'
  result: string
  resultJsTokens: {
    tokens: any[]
    error?: any
  }
} {
  const resultJsTokens = _stripLiteralJsTokens(code, options)
  if (!resultJsTokens.error) {
    return {
      mode: 'js-tokens',
      result: resultJsTokens.result,
      resultJsTokens,
    }
  }
  return {
    mode: 'regex',
    result: stripLiteralRegex(resultJsTokens.result + code.slice(resultJsTokens.result.length), options),
    resultJsTokens,
  }
}
