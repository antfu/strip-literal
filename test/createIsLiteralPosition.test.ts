/* eslint-disable no-template-curly-in-string */
import { expect, test } from 'vitest'
import { createIsLiteralPositionAcorn } from '../src'

function execute(code: string) {
  const isLiteralPosition = createIsLiteralPositionAcorn(code)

  const positions = new Array(code.length)
    .fill(0)
    .map((_, i) => i)
  const result = positions
    .map((pos) => {
      if (code[pos] === '\n')
        return isLiteralPosition(pos) ? '+\n' : '\n'
      return isLiteralPosition(pos) ? '*' : code[pos]
    })
    .join('')

  return result
}

test('works', () => {
  expect(execute(`
const a = 0
  `)).toMatchSnapshot()
  expect(execute(`
// a
const a = 0
  `)).toMatchSnapshot()
  expect(execute(`
/* a */
const a = 0
  `)).toMatchSnapshot()
  expect(execute(`
/*
  a
*/
const a = 0
  `)).toMatchSnapshot()
  expect(execute(`
const a = 'a'
  `)).toMatchSnapshot()
  expect(execute(`
const a = "a"
  `)).toMatchSnapshot()
  expect(execute(`
const a = \`c\${b}\`
  `)).toMatchSnapshot()
})

test('template string nested', () => {
  expect(execute(
    '`aa${a + `a`}aa`',
  )).toMatchSnapshot()
})
