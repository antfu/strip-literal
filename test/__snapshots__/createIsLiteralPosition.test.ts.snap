// Vitest Snapshot v1

exports[`template string nested 1`] = `"\`**\${a + \`*\`}**\`"`;

exports[`works 1`] = `
"
const a = 0
  "
`;

exports[`works 2`] = `
"
****
const a = 0
  "
`;

exports[`works 3`] = `
"
*******
const a = 0
  "
`;

exports[`works 4`] = `
"
**+
***+
**
const a = 0
  "
`;

exports[`works 5`] = `
"
const a = '*'
  "
`;

exports[`works 6`] = `
"
const a = \\"*\\"
  "
`;

exports[`works 7`] = `
"
const a = \`*\${b}\`
  "
`;
